File: kotlin/src/main/kotlin/io/dpmp/protocol/Codec.kt
package io.dpmp.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * DPMP 二进制编解码。
 *
 * UDP-RTP 包头（大端）：
 *     [1B type][4B seq][4B ack][2B length][1B reserved] = 12 字节
 *
 * 与 Python 端 dpmp/protocol/codec.py 逐字节一致。
 */
object Codec {

    /** 12 字节包头：!B I I H B */
    const val HEADER_SIZE = C.RTP_HEADER_SIZE

    /** 解析后的包头。 */
    data class Header(
        val type: Int,
        val seq: Long,
        val ack: Long,
        val length: Int,
        val reserved: Int
    )

    // ---------- 包头 ----------

    /** 打包 12 字节 UDP-RTP 包头。 */
    fun packHeader(
        ptype: Int, seq: Long = 0L, ack: Long = 0L,
        length: Int = 0, reserved: Int = 0
    ): ByteArray {
        val buf = ByteBuffer.allocate(HEADER_SIZE).order(ByteOrder.BIG_ENDIAN)
        buf.put((ptype and 0xFF).toByte())
        buf.putInt((seq and 0xFFFFFFFFL).toInt())
        buf.putInt((ack and 0xFFFFFFFFL).toInt())
        buf.putShort((length and 0xFFFF).toShort())
        buf.put((reserved and 0xFF).toByte())
        return buf.array()
    }

    /** 解析 12 字节包头；不足 12 字节返回 null。 */
    fun unpackHeader(data: ByteArray): Header? {
        if (data.size < HEADER_SIZE) return null
        val buf = ByteBuffer.wrap(data, 0, HEADER_SIZE).order(ByteOrder.BIG_ENDIAN)
        val type = buf.get().toInt() and 0xFF
        val seq = buf.int.toLong() and 0xFFFFFFFFL
        val ack = buf.int.toLong() and 0xFFFFFFFFL
        val length = buf.short.toInt() and 0xFFFF
        val reserved = buf.get().toInt() and 0xFF
        return Header(type, seq, ack, length, reserved)
    }

    /** 打包一个完整报文（头 + payload）。 */
    fun packPacket(ptype: Int, seq: Long, ack: Long, payload: ByteArray = ByteArray(0)): ByteArray {
        val hdr = packHeader(ptype, seq, ack, payload.size, 0)
        val out = ByteArray(hdr.size + payload.size)
        System.arraycopy(hdr, 0, out, 0, hdr.size)
        System.arraycopy(payload, 0, out, hdr.size, payload.size)
        return out
    }

    // ---------- SYNC_REQ ----------
    // !QB : t1(ms), proto_ver

    fun packSyncReq(t1Ms: Long, protoVer: Int = C.RTP_PROTO_VER): ByteArray {
        val buf = ByteBuffer.allocate(9).order(ByteOrder.BIG_ENDIAN)
        buf.putLong(t1Ms)
        buf.put((protoVer and 0xFF).toByte())
        return buf.array()
    }

    /** 返回 [t1Ms, protoVer]。 */
    fun unpackSyncReq(payload: ByteArray): Pair<Long, Int> {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
        val t1 = buf.long
        val ver = buf.get().toInt() and 0xFF
        return t1 to ver
    }

    // ---------- SYNC_ACK ----------
    // !QQQB : t1, t2, t3, proto_ver

    fun packSyncAck(t1: Long, t2: Long, t3: Long, protoVer: Int = C.RTP_PROTO_VER): ByteArray {
        val buf = ByteBuffer.allocate(25).order(ByteOrder.BIG_ENDIAN)
        buf.putLong(t1); buf.putLong(t2); buf.putLong(t3)
        buf.put((protoVer and 0xFF).toByte())
        return buf.array()
    }

    /** 返回 [t1, t2, t3, protoVer]。 */
    fun unpackSyncAck(payload: ByteArray): LongArray {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
        val t1 = buf.long; val t2 = buf.long; val t3 = buf.long
        val ver = (buf.get().toInt() and 0xFF).toLong()
        return longArrayOf(t1, t2, t3, ver)
    }

    // ---------- SYNC_COMMIT ----------
    // v2 !QQ : t_go_r, commit_id
    // v1 !Q  : delta_ms

    fun packSyncCommitV2(tGoR: Long, commitId: Long): ByteArray {
        val buf = ByteBuffer.allocate(16).order(ByteOrder.BIG_ENDIAN)
        buf.putLong(tGoR); buf.putLong(commitId)
        return buf.array()
    }

    fun unpackSyncCommitV2(payload: ByteArray): Pair<Long, Long> {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
        return buf.long to buf.long
    }

    fun packSyncCommitV1(deltaMs: Long): ByteArray {
        val buf = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN)
        buf.putLong(deltaMs)
        return buf.array()
    }

    fun unpackSyncCommitV1(payload: ByteArray): Long {
        return ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN).long
    }

    // ---------- PUNCH_ROUND ----------
    // !QQ : round_no, t_go_r

    fun packPunchRound(roundNo: Long, tGoR: Long): ByteArray {
        val buf = ByteBuffer.allocate(16).order(ByteOrder.BIG_ENDIAN)
        buf.putLong(roundNo); buf.putLong(tGoR)
        return buf.array()
    }

    fun unpackPunchRound(payload: ByteArray): Pair<Long, Long> {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
        return buf.long to buf.long
    }
}

File: dpmp/protocol/codec.py
# -*- coding: utf-8 -*-
"""DPMP 二进制编解码。

UDP-RTP 包头（大端）：
    [1B type][4B seq][4B ack][2B length][1B reserved] = 12 字节
"""

import struct
from typing import Optional, Tuple

from . import constants as C

# UDP-RTP 12 字节包头：!B I I H B
RTP_HEADER = struct.Struct("!BIIHB")

# 各控制报文的 payload 结构
_PAYLOAD_SYNC_REQ = struct.Struct("!QB")      # t1(ms), proto_ver
_PAYLOAD_SYNC_ACK = struct.Struct("!QQQB")    # t1, t2, t3, proto_ver
_PAYLOAD_SYNC_COMMIT_V2 = struct.Struct("!QQ")  # t_go_r, commit_id
_PAYLOAD_SYNC_COMMIT_V1 = struct.Struct("!Q")   # delta_ms
_PAYLOAD_PUNCH_ROUND = struct.Struct("!QQ")     # round_no, t_go_r


def pack_header(ptype: int, seq: int = 0, ack: int = 0,
                length: int = 0, reserved: int = 0) -> bytes:
    """打包 12 字节 UDP-RTP 包头。"""
    return RTP_HEADER.pack(ptype, seq, ack, length, reserved)


def unpack_header(data: bytes) -> Optional[Tuple[int, int, int, int, int]]:
    """解析 12 字节包头，返回 (type, seq, ack, length, reserved)。

    数据不足 12 字节时返回 None。
    """
    if len(data) < C.RTP_HEADER_SIZE:
        return None
    try:
        return RTP_HEADER.unpack(data[:C.RTP_HEADER_SIZE])
    except struct.error:
        return None


def pack_packet(ptype: int, seq: int, ack: int, payload: bytes = b"") -> bytes:
    """打包一个完整的 UDP-RTP 报文（头 + payload）。"""
    return pack_header(ptype, seq, ack, len(payload)) + payload


# ---------- SYNC_REQ ----------

def pack_sync_req(t1_ms: int, proto_ver: int = C.RTP_PROTO_VER) -> bytes:
    return _PAYLOAD_SYNC_REQ.pack(t1_ms, proto_ver)


def unpack_sync_req(payload: bytes) -> Tuple[int, int]:
    return _PAYLOAD_SYNC_REQ.unpack(payload)


# ---------- SYNC_ACK ----------

def pack_sync_ack(t1: int, t2: int, t3: int,
                  proto_ver: int = C.RTP_PROTO_VER) -> bytes:
    return _PAYLOAD_SYNC_ACK.pack(t1, t2, t3, proto_ver)


def unpack_sync_ack(payload: bytes) -> Tuple[int, int, int, int]:
    return _PAYLOAD_SYNC_ACK.unpack(payload)


# ---------- SYNC_COMMIT ----------

def pack_sync_commit_v2(t_go_r: int, commit_id: int) -> bytes:
    return _PAYLOAD_SYNC_COMMIT_V2.pack(t_go_r, commit_id)


def unpack_sync_commit_v2(payload: bytes) -> Tuple[int, int]:
    return _PAYLOAD_SYNC_COMMIT_V2.unpack(payload)


def pack_sync_commit_v1(delta_ms: int) -> bytes:
    return _PAYLOAD_SYNC_COMMIT_V1.pack(delta_ms)


def unpack_sync_commit_v1(payload: bytes) -> Tuple[int]:
    return _PAYLOAD_SYNC_COMMIT_V1.unpack(payload)


# ---------- PUNCH_ROUND ----------

def pack_punch_round(round_no: int, t_go_r: int) -> bytes:
    return _PAYLOAD_PUNCH_ROUND.pack(round_no, t_go_r)


def unpack_punch_round(payload: bytes) -> Tuple[int, int]:
    return _PAYLOAD_PUNCH_ROUND.unpack(payload)
